docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,924 @@
+for same author/timestamp/realm_id/key_index.
+
+Each key in the Keys-bundle must be checked against the canary in `RealmKeyRotationCertificate`
+that introduced it.
+
+At any time, `realm_share`/`vlob_read`/`vlob_create` can return `bad_key_index` if the client
+consider the wrong key to be the last one. In this case the client simply forgets its realm
+keys-bundle and re-do the fetch operation.
+
+### 6.4 - Keys-bundle self-healing
+
+A keys-bundle can be corrupted in multiple ways:
+
+1. Invalid data: bad signature, bad encryption, bad serialization
+2. Additional keys
+3. Missing keys
+
+On top of that, a faulty key rotation may end up with a key rotation certificate and an
+invalid corresponding key (i.e. the encryption key is forever lost). However no matter
+what the key-bundle must contains as much key as there is key rotation certificates.
+
+Hence a keys-bundle containing tempered key is not considered corrupted: as long
+the document is properly signed and contains the right number of keys, it is
+considered valid. And, of course, any key in the key bundle must first be checked
+against the canary contained in it corresponding key rotation certificate before
+actually be used.
+
+Hence the self-healing process consist of recursively fetching the previous key rotation
+certificate until either one of them is valid, or there is no more available for us (i.e.
+the key bundle was already broken when the realm got shared with us).
+
+In any case, client should dispatch internally an event to inform of corruption
+(including the author of the faulty key rotation) and self-healing outcome. This event
+typically ends up as a warning logged and gets displayed in the GUI.
+
+> **Note:**
+>
+> We cannot simply get the keys-bundle that originally created a given key given
+> the realm might have been shared with the user later on.
+
+### 6.5 - Automatic key rotation
+
+For automatic key rotation, a new "key rotation manager" is to be introduced. Its function
+is very simple:
+
+1. Wait for new `UserRevokedCertificate`, `RealmRoleCertificate` and
+   `RealmKeyRotationCertificate` certificates
+2. For `UserRevokedCertificate`, call `realm_unshare` for each realm we are owner (and
+   the revoked user is part of)
+3. For `RealmRoleCertificate`, if the role is None and we are owner of the realm:
+    1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
+       already waiting for this realm)
+    2. Call `realm_rotate_key`
+4. For `RealmKeyRotationCertificate` cancel the wait from step 3.1 if any
+
+The wait at step 3.1 is here for two reasons:
+
+- Give priority to the author of the certificates to do the key rotation. This is
+  typically needed if he is doing un-sharing batch (and hence will do the key rotation
+  only once all un-sharing have been done)
+- Avoid all owners to send key rotation command at the same time to the server.
```

Comment:
Given 2 owner that send a key-rotation command at the "same" time (we are unlucky, or not :thinking:), what would append ?

Is it first arrived, first served ?

Reply:
> Is it first arrived, first served ?

yes, the server checks the key rotation index against the last one in database and returns a `bad_key_index` status